glean-core/ffi/src/lib.rs

```diff
@@ -198,16 +196,51 @@ pub extern "C" fn glean_enable_logging() {
     }
 }
 
-#[no_mangle]
-pub extern "C" fn glean_initialize(
-    data_dir: FfiStr,
-    application_id: FfiStr,
+#[repr(C)]
+pub struct FfiConfiguration<'a> {
+    data_dir: FfiStr<'a>,
+    package_name: FfiStr<'a>,
     upload_enabled: u8,
-) -> u64 {
+    max_events: Option<&'a i32>,
+}
+
+use std::convert::TryFrom;
+impl TryFrom<&FfiConfiguration<'_>> for glean_core::Configuration {
+    type Error = glean_core::Error;
+    fn try_from(cfg: &FfiConfiguration) -> Result<Self, Self::Error> {
+        let data_path = cfg
+            .data_dir
+            .as_opt_str()
+            .map(|s| s.to_string())
+            .ok_or_else(glean_core::Error::utf8_error)?;
+        let application_id = cfg
+            .package_name
+            .as_opt_str()
+            .map(|s| s.to_string())
+            .ok_or_else(glean_core::Error::utf8_error)?;
+        let upload_enabled = cfg.upload_enabled != 0;
+        let max_events = cfg.max_events.map(|m| *m as usize);
+
+        Ok(Self {
+            upload_enabled,
+            data_path,
+            application_id,
+            max_events,
+        })
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn glean_initialize(cfg: *const FfiConfiguration) -> u64 {
+    assert!(!cfg.is_null());
+
     GLEAN.insert_with_log(|| {
-        let data_dir = data_dir.into_string();
-        let application_id = application_id.into_string();
-        let glean = Glean::new(&data_dir, &application_id, upload_enabled != 0)?;
+        // We can create a reference to the FfiConfiguration strut:
+        // 1. We did a null check
+        // 2. We're not holding on to it beyond this function,
+        //    we copy out all data when needed
```

Comment:
nit: leading whitespaces and missing trailing dot

Reply:
Whitespace intended, so it lines up with the line it belongs to.